### PR TITLE
tests: Use test framework utils where possible

### LINKED ISSUE #23119: test: replace bare asserts with assertion helpers (assert_equal() etc.)
#23117 replaced asserts with the test framework's internal helpers (see https://github.com/bitcoin/bitcoin/blob/dccf3d25f9e78909eb7b3143e89a7c87fac25ab5/test/functional/test_framework/util.py#L47-L59) for a single test, in order to see the expected and failed values if such an assertion fails. The same should be done for all the remaining tests. Potential candidates can be found via
```
$ cd ./test/functional
$ git grep assert.*==
```

#### Useful skills:

basic Python3 knowledge

#### Want to work on this issue?

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
This PR removes when it is possible the native python assert, and use the test utils function, like `assert_equal`, `assert_greater_than`, `assert_greater_than_or_equal`.

### Update

In addition, this PR includes a new operator `assert_not_equal`, which IMO it is useful to test other conditions and add the error message as a parameter of the function to give more flexibility.

Fixes https://github.com/bitcoin/bitcoin/issues/23119

Edit laanwj: removed HTML comment from post

### CODE REVIEW

File: test/functional/feature_backwards_compatibility.py [hunk 1]
```diff
@@ -123,52 +124,52 @@ def run_test(self):
         wallet = node_v18.get_wallet_rpc("w1_v18")
         info = wallet.getwalletinfo()
         assert info['private_keys_enabled']
-        assert info['keypoolsize'] > 0
+        assert_greater_than(info['keypoolsize'] > 0)
```
Comment (meshcollider) [inline test/functional/feature_backwards_compatibility.py:127]: error here, `,` instead of `>`

**Comment by meshcollider** [comment]
Thanks for contributing and helping out @vincenzopalazzo!

File: test/functional/feature_backwards_compatibility.py (hunk 1 above)
Comment (vincenzopalazzo) [reply test/functional/feature_backwards_compatibility.py]: Ops, thanks to catch my error @meshcollider!

**Comment by brunoerg** [comment]
Concept ACK

**Comment by brunoerg** [comment]
Will it be created new PRs to update other tests or should we concentrate all the changes here?

**Comment by vincenzopalazzo** [comment]
>Will it be created new PRs to update other tests or should we concentrate all the changes here?

The plan is to divide the work into small chunks to make the price of PR review easy and fast.

**Comment by brunoerg** [comment]
> > Will it be created new PRs to update other tests or should we concentrate all the changes here?
> 
> The plan is to divide the work into small chunks to make the price of PR review easy and fast.

Cool, thanks!

**Comment by brunoerg** [comment]
I just created #23135 to help on it

**Comment by vincenzopalazzo** [comment]
>I just created #23135 to help on it

Thanks, the important think it that we don't make the double work :-)

**Comment by michaelfolkson** [comment]
Concept ACK

I think generally splitting across multiple pull requests works best for the monster sub-projects ([AssumeUTXO](https://github.com/bitcoin/bitcoin/projects/11), [deglobalizing ChainstateManager](https://github.com/bitcoin/bitcoin/pull/20158), [Process Separation](https://github.com/bitcoin/bitcoin/projects/10) etc).

In this case splitting across multiple commits within the same PR would probably have been sufficient to make it easily reviewable but not a big deal and if it encourages multiple contributors to self-organize around completing it then that's great.

File: test/functional/test_framework/util.py [hunk 2]
```diff
@@ -50,6 +50,11 @@ def assert_equal(thing1, thing2, *args):
         raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
 
 
+def assert_not_equal(thing1, thing2, *args):
+    if thing1 == thing2 or any(thing1 != arg for arg in args):
```
Comment (kiminuo) [inline test/functional/test_framework/util.py:54]: Should it possibly be:

```suggestion
    if thing1 == thing2 or any(thing1 == arg for arg in args):
```

?

Comment (vincenzopalazzo) [reply test/functional/test_framework/util.py]: Good catch @kiminuo thanks you!

**Comment by vincenzopalazzo** [comment]
Thanks @michaelfolkson,

>In this case splitting across multiple commits within the same PR would probably have been sufficient to make it easily reviewable but not a big deal and if it encourages multiple contributors to self-organize around completing it then that's great.

It sounds great to me, I will go with this solution. Thansk!

**Comment by laanwj** [comment]
Concept ACK 9edb3cc5d08caaf61ff00c49a698956f4ddaba08, looked over the commits briefly and a) didn't find any mistakes b) checked the changes are contained to where they should be

though I haven't reviewed the huge list of changes in detail so can't warrant calling it a full code review ACK.

**Comment by vincenzopalazzo** [comment]
>though I haven't reviewed the huge list of changes in detail so can't warrant calling it a full code review ACK.

Sorry about that, I should add some tips on the PR. However, there is only a critical change in the test utils framework, now all the assert support the error message as the parameter this only because we can support also custom message defined with `assert False, "always false"`

**Comment by laanwj** [comment]
> Sorry about that, I should add some tips on the PR. 

I think you misunderstood: It was not a comment on your work. I meant that I can't call what I did a full code review ACK. I should look in more detail, though generally getting the type of `assert` wrong will usually just result in the tests failing, so I'm reasonably confident in the correctness of the changes.

**Comment by michaelfolkson** [comment]
> I think generally splitting across multiple pull requests works best for the monster sub-projects (AssumeUTXO, deglobalizing ChainstateManager, Process Separation etc).

I know I've previously said this (so apologies if this wasn't helpful at the time) but I think the PR has got sufficiently big that a PR or two could maybe be spun out of this one to make it easier to review.

I also think the structure of commits could be improved, having a separate commit for each file doesn't make much sense to me. Also the commit messages aren't particularly informative for reviewers.

Commit messages such as "Define new assert functions in test utils", "Replace assert with assert_equal across functional test suite" and "Use assert_greater_than instead of greater than asserts" would be more helpful than "filename.py: Use test framework".

Reviewers can see what files are changed without looking at commit messages, what they want to see with the commit messages is what the commit is doing.

**Comment by CerberusVT** [comment]
> > I think generally splitting across multiple pull requests works best for the monster sub-projects (AssumeUTXO, deglobalizing ChainstateManager, Process Separation etc).
> 
> I know I've previously said this (so apologies if this wasn't helpful at the time) but I think the PR has got sufficiently big that a PR or two could maybe be spun out of this one to make it easier to review.
> 
> I also think the structure of commits could be improved, having a separate commit for each file doesn't make much sense to me. Also the commit messages aren't particularly informative for reviewers.
> 
> Commit messages such as "Define new assert functions in test utils", "Replace assert with assert_equal across functional test suite" and "Use assert_greater_than instead of greater than asserts" would be more helpful than "filename.py: Use test framework".
> 
> Reviewers can see what files are changed without looking at commit messages, what they want to see with the commit messages is what the commit is doing.

Hey, I agree with you.

Maybe this pull request could be splitted into three different ones.
For example: one pull request per each assertion type to refactor `assert_equal`, `assert_greater_than` and `assert_greater_than_or_equal`

I'd would like to help with the reviews!

**Comment by vincenzopalazzo** [comment]
>Maybe this pull request could be splitted into three different ones.
For example: one pull request per each assertion type to refactor assert_equal, assert_greater_than and

Each following step required the same trade-off with only personal preference, if you split the PR in more PRs, you have more work to do, and so on.

The change is not only in the change of the assert tp assert_* but this PR includes also some change inside the test framework, now the method accepts also an optional parameter `err_msg`

This required a big pr that include this change too.

In conclusion, I agree that small PR can be useful for other people, but at the end of the day, there is no real improvement inside the velocity of review.

This takes time, as with all types of refactoring. This time is chosen this way but the next time we can do in another way that we can set up before finish the work. Back in October, I chatted on IRC to choose which method is the best, so this is not my own choice.

File: test/functional/feature_segwit.py [hunk 3]
```diff
@@ -44,6 +44,7 @@
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import (
     assert_equal,
+    assert_not_equal,
     assert_greater_than_or_equal,
     assert_is_hex_string,
```
Comment (niVelion) [inline test/functional/feature_segwit.py:49]: Nit. This file has alphabetical imports.

```suggestion
    assert_greater_than_or_equal,
    assert_is_hex_string,
    assert_not_equal,
```

File: test/functional/feature_taproot.py [hunk 4]
```diff
@@ -91,7 +91,14 @@
...
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_raises_rpc_error, assert_equal
+from test_framework.util import (
+    assert_raises_rpc_error,
+    assert_equal,
+    assert_not_equal,
+    assert_true,
+    assert_greater_than,
+    assert_greater_than_or_equal,
+)
```
Comment (niVelion) [inline test/functional/feature_taproot.py:101]: Nit.

```suggestion
from test_framework.util import (
    assert_raises_rpc_error,
    assert_equal,
    assert_not_equal,
    assert_true,
    assert_greater_than,
    assert_greater_than_or_equal,
)
```

File: test/functional/interface_zmq.py [hunk 5]
```diff
@@ -511,16 +513,16 @@ def test_mempool_sync(self):
...
                     else:
                         raise Exception(f"WARNING: txhash has unexpected mempool sequence value: {mempool_sequence} vs expected {expected_sequence}")
             if label == "A":
-                assert hash_str not in mempool_view
+                assert_true(hash_str not in mempool_view)
                 mempool_view.add(hash_str)
                 expected_sequence = mempool_sequence + 1
             elif label == "R":
-                assert hash_str in mempool_view
+                assert_true(hash_str in mempool_view)
```
Comment (MarcoFalke) [inline test/functional/interface_zmq.py:525]: what is the motivation to change this?

File: test/functional/wallet_txn_doublespend.py [hunk 6]
```diff
@@ -46,7 +46,7 @@ def run_test(self):
         # blockchain sync later in the test when nodes are connected, due to
         # timing issues.
         for n in self.nodes:
-            assert n.getblockchaininfo()["initialblockdownload"] == False
+            assert_equal(n.getblockchaininfo()["initialblockdownload"], False)
```
Comment (MarcoFalke) [inline test/functional/wallet_txn_doublespend.py:49]: Seems there are too many ways to say the same thing. I think we shouldn't be adding  even more.

For boolean values, if the assertion fails, it is clear what value the bool was. There is no need to print the value and a constant.

**Review by niVelion — COMMENTED** [review]
Thoroughly reviewed and just for good measure, checked out compiled and ran `test/functional/test_runner.py --extended`.

File: test/functional/wallet_txn_doublespend.py (hunk 6 above)
Comment (vincenzopalazzo) [reply test/functional/wallet_txn_doublespend.py:49]: >Seems there are too many ways to say the same thing. I think we shouldn't be adding even more.

Right, in this case, I use the `assert_equal` but I could use `assert_true`, and the reason that I used this is only for consistency across error messages and source code. However, if you like I can restore the pure `assert`

File: test/functional/interface_zmq.py (hunk 5 above)
Comment (vincenzopalazzo) [reply test/functional/interface_zmq.py:525]: The only reason is to maintain consistency between error code format and code style. I usually try to avoid mixing different types of assert (from library and native assert) in the tests

File: test/functional/feature_taproot.py [hunk 7]
```diff
@@ -585,7 +592,7 @@ def bitflipper(expr):
     """Return a callable that evaluates expr and returns it with a random bitflip."""
     def fn(ctx):
         sub = deep_eval(ctx, expr)
-        assert isinstance(sub, bytes)
+        assert_true(isinstance(sub, bytes), err_msg="It is not a bytes instance")
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply test/functional/feature_taproot.py:595]: You have right! I will change the error message, but here I really have no idea what is the correct error message to add. Any suggestion?

File: test/functional/p2p_compactblocks.py [hunk 8]
```diff
@@ -729,7 +731,7 @@ def test_end_to_end_block_relay(self, listeners):
     # but invalid transactions.
     def test_invalid_tx_in_compactblock(self, test_node, use_segwit=True):
         node = self.nodes[0]
-        assert len(self.utxos)
+        assert_true(len(self.utxos) > 0)
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply test/functional/p2p_compactblocks.py:734]: I make this choice for consistency, across error messages and assert convention. I usually try to avoid missing pure assert and assert that came from utils code like the bitcoin test framework.

But I admit that in this case, it would be better `assert_not_empity(self.utxos)`

However, if you disagree I would restore the previous status with the pure assert

File: test/functional/interface_zmq.py (hunk 5 above)
Comment (MarcoFalke) [reply test/functional/interface_zmq.py:525]: It is fine if you do this personally in your own projects, but for bitcoin-core this is impossible to maintain. Someone is going to break the consistency at some point. We can't handle this if there are floods of follow-up pull requests. Also, if there is a linter to make lib assert illegal, it puts an unnecessary burden on developers for style reasons.

**Comment by kallewoof** [comment]
I don't see why splitting into one commit per file does anything to help review. The contrary, I'd say. The UI already splits per file for you.

**Comment by vincenzopalazzo** [comment]
>I don't see why splitting into one commit per file does anything to help review. The contrary, I'd say. The UI already splits per file for you.

When I start to work on that I had the opposite advice, I think this could be a personal choice, I'm sorry if it makes things worse for you.

I'm fine with this splitting because each commit has only a personal portion of all the change, and usually, the UI starts to be very slow for a lot of changes. In addition, for disability reason, I don't use the UI to review PR, but only to leave a comment, so I considered this approach more accessible too

File: test/functional/interface_zmq.py (hunk 5 above)
Comment (vincenzopalazzo) [reply test/functional/interface_zmq.py]: Mh I agree with the difficulty to add a change of style at this point on Bitcoin core, I only propose one way here and I'm super happy to revert the change and to have this great exchange of idea

File: test/functional/wallet_txn_doublespend.py (hunk 6 above)
Comment (michaelfolkson) [reply test/functional/wallet_txn_doublespend.py:49]: @MarcoFalke: What is your preference for what @vincenzopalazzo does here?

**Comment by michaelfolkson** [comment]
>> I don't see why splitting into one commit per file does anything to help review. The contrary, I'd say. The UI already splits per file for you.

> When I start to work on that I had the opposite advice, I think this could be a personal choice, I'm sorry if it makes things worse for you.

I don't think anyone suggested that it should be one commit per file right? I suggested splitting across commits rather than different PRs but not giving each file a separate commit.

I suggested this:

> I also think the structure of commits could be improved, having a separate commit for each file doesn't make much sense to me. Also the commit messages aren't particularly informative for reviewers.
>
> Commit messages such as "Define new assert functions in test utils", "Replace assert with assert_equal across functional test suite" and "Use assert_greater_than instead of greater than asserts" would be more helpful than "filename.py: Use test framework".
>
> Reviewers can see what files are changed without looking at commit messages, what they want to see with the commit messages is what the commit is doing.

I personally think this isn't serious enough (others can assess) that it means it is not yet in a merge-able state but yeah I agree with Kalle that splitting into one commit per file isn't optimal.

File: test/functional/feature_taproot.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (laanwj) [reply test/functional/feature_taproot.py:596]: Asserting prints the line where the error happens, right? The expression `isinstance(sub, bytes)` says it all, I don't think any more information needs to be added here.

File: test/functional/test_framework/util.py [hunk 9]
```diff
@@ -66,6 +66,11 @@ def assert_greater_than_or_equal(thing1, thing2):
         raise AssertionError("%s < %s" % (str(thing1), str(thing2)))
 
 
+def assert_true(thing, message = None):
+    if thing is not True:
+        msg = "%s it is not True" if message is None else message
```
Comment (laanwj) [inline test/functional/test_framework/util.py]: `it` needs to be removed from this sentence, also `% str(thing)` needs to be used to perform formatting so that the value of `thing` ends up in the message.

**Comment by vincenzopalazzo** [comment]
Trivial rebase!

File: test/functional/wallet_txn_doublespend.py (hunk 6 above)
Comment (stickies-v) [reply test/functional/wallet_txn_doublespend.py:49]: > For boolean values, if the assertion fails, it is clear what value the bool was. There is no need to print the value and a constant.

I'd agree in a statically typed environment, but in Python that doesn't really hold? I think there's merit in always printing the value, to more quickly debug cases where e.g. the type of a response changes or becomes None/empty string?

**Comment by fanquake** [comment]
@vincenzopalazzo Given #25732, did you want to leave this one open as well, or close for now?

**Comment by vincenzopalazzo** [comment]
better close this and hope that someone will finish the job :)
